sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -85,77 +232,81 @@ case class CachedRDDBuilder(
-          }.toArray
-
-          var rowCount = 0
-          var totalSize = 0L
-          while (rowIterator.hasNext && rowCount < batchSize
-            && totalSize < ColumnBuilder.MAX_BATCH_SIZE_IN_BYTE) {
-            val row = rowIterator.next()
-
-            // Added for SPARK-6082. This assertion can be useful for scenarios when something
-            // like Hive TRANSFORM is used. The external data generation script used in TRANSFORM
-            // may result malformed rows, causing ArrayIndexOutOfBoundsException, which is somewhat
-            // hard to decipher.
-            assert(
-              row.numFields == columnBuilders.length,
-              s"Row column number mismatch, expected ${output.size} columns, " +
-                s"but got ${row.numFields}." +
-                s"\nRow content: $row")
-
-            var i = 0
-            totalSize = 0
-            while (i < row.numFields) {
-              columnBuilders(i).appendFrom(row, i)
-              totalSize += columnBuilders(i).columnStats.sizeInBytes
-              i += 1
-            }
-            rowCount += 1
-          }
-
-          sizeInBytesStats.add(totalSize)
-          rowCountStats.add(rowCount)
-
-          val stats = InternalRow.fromSeq(
-            columnBuilders.flatMap(_.columnStats.collectedStatistics).toSeq)
-          CachedBatch(rowCount, columnBuilders.map { builder =>
-            JavaUtils.bufferToArray(builder.build())
-          }, stats)
-        }
-
-        def hasNext: Boolean = rowIterator.hasNext
-      }
+    val cb = if (cachedPlan.supportsColumnar) {
+      serializer.convertForCacheColumnar(cachedPlan.executeColumnar(),
+        cachedPlan.output,
+        storageLevel,
+        cachedPlan.conf)
+    } else {
+      serializer.convertForCache(cachedPlan.execute(),
+        cachedPlan.output,
+        storageLevel,
+        cachedPlan.conf)
+    }
+    val cached = cb.map { batch =>
+      sizeInBytesStats.add(batch.sizeInBytes)
+      rowCountStats.add(batch.numRows)
+      batch
     }.persist(storageLevel)
-
     cached.setName(cachedName)
     cached
   }
 }
 
 object InMemoryRelation {
 
+  private[this] var ser: Option[CachedBatchSerializer] = None
+  private[this] def getSerializer(sqlConf: SQLConf): CachedBatchSerializer = synchronized {
+    if (ser.isEmpty) {
+      val serName = sqlConf.getConf(StaticSQLConf.SPARK_CACHE_SERIALIZER)
+      val serClass = Utils.classForName(serName)
+      val instance = serClass.getConstructor().newInstance().asInstanceOf[CachedBatchSerializer]
+      ser = Some(instance)
+    }
+    ser.get
+  }
+
   def apply(
-      useCompression: Boolean,
-      batchSize: Int,
+      storageLevel: StorageLevel,
+      qe: QueryExecution,
+      tableName: Option[String]): InMemoryRelation = {
+    val optimizedPlan = qe.optimizedPlan
+    val serializer = getSerializer(optimizedPlan.conf)
+    val child = if (serializer.supportsColumnarInput(optimizedPlan.output)) {
```

Comment:
This is the part I want to clarify.

If the serializer can't support columnar input, it's easy and we just use the original query plan.

If the serializer supports columnar input, what shall we do?
1. if the serializer can't support row-based input (is it possible?), we have to add a row -> columnar operator.
2. If the serializer also supports row-based input, but the query plan is row-based (no parquet scan), does it worth to add a row -> columnar operator?

Reply:
I picked row based input as a minimum requirement because the majority of the time the output of the plan will be row based. I also wanted to avoid unnecessary transitions from one data format to another. Each time one tries to change the format of the data, row to column, column to row, or even column of one type to column of another there is a cost involved. As a rule of thumb it is ideal to reduce those transformations when ever possible. So in that respect columnar input support is a performance optimization in the rare case that columnar input is a possibility.

Reply:
That makes sense.

Then my question is: how do we know a plan can produce columnar data? I see you added quite a bit of changes for it, but I've figured it out yet.

I expect something like
```
plan.isInstanceOf[WholeStageCodegen] && plan.children.head.isInstanceOf[ColumnarToRow]
```

Maybe I missed something. Can you elaborate it a bit more?

Reply:
Sorry I didn't explain.  I wanted to avoid reflection if possible, because our GPU plugin can replace `ColumnarToRow` with a version that also pulls data back from the `GPU`, and with (SPARK-32334)[https://issues.apache.org/jira/browse/SPARK-32334] there is the possibility around that more things related to columnar transitions might change. So instead of trying to pick apart the incoming `SparkPlan` to remove the `ColumnarToRow`, I added an API to `QueryExecution` so it can generate a plan without the transition in it.

If you would prefer me to try and pick apart the plan I can do that instead. I just thought this was a cleaner approach.

Reply:
I like SPARK-32334 , maybe we should do that first.

Say if we have a trait/abstract class for `ColumnarToRow`, then we can safely play `isInstanceOf` here, as your customized version of columnar to row operator should also extend the trait/abstract class.

Reply:
BTW if we want to unblock this PR, I'd suggest we do `isInstanceOf[ColumnarToRow]` first, and waiting for SPARK-32334 to extend it. I feel a bit worried about adding hacks to `QueryExecution`, as it's very critical.

Reply:
I'll switch over to doing `isInstanceOf[ColumnarToRow]` just to get this unblocked.

I don't consider the change I made a hack. It is a relatively small and focused change. It also will result in less implicit coupling between the cache code and the columnar planning code.

I like SPARK-32334 too, but for what I have proposed so far it really only covers transformations to arrow. If you have ideas on how you want it to impact these changes please add some comments there so we can discuss it and get a better idea of how it should all fit together.

Reply:
or, can we do the check with
```
!plan.supportsColumnar && plan.child.supportsColumnar
```

Reply:
I switched the patch to do what you requested.